### PR TITLE
[WIP] Use haskell-lsp with VFS race condition fix

### PR DESCRIPTION
**Pull request by alanz**
Use work @mpickering has done to get rid of race conditions in VFS

Do not merge, this is a test branch, to confirm operation prior to publishing new `haskell-lsp` and `lsp-test` modules.

### CODE REVIEW

File: src/Haskell/Ide/Engine/Transport/LspStdio.hs [hunk 1]
```diff
@@ -966,16 +967,34 @@ syncOptions = J.TextDocumentSyncOptions
...
+      -- may list out every specific kind they provide.
+      -- , codeActionKinds                  :: Maybe [J.CodeActionKind]
+
+      -- The list of characters that triggers on type formatting.
+      -- If you set `documentOnTypeFormattingHandler`, you **must** set this.
+      , Core.documentOnTypeFormattingTriggerCharacters = nonEmpty []
+
+      -- The commands to be executed on the server.
+      -- If you set `executeCommandHandler`, you **must** set this.
+      , Core.executeCommandCommands           = Just commandIds
```
Comment (alanz) [inline src/Haskell/Ide/Engine/Transport/LspStdio.hs]: I am not sure if these are set up correctly, please provide input.

File: hie-plugin-api/Haskell/Ide/Engine/PluginUtils.hs [hunk 2]
```diff
@@ -275,7 +275,7 @@ readVFS :: (MonadIde m, MonadIO m) => Uri -> m (Maybe T.Text)
 readVFS uri = do
   mvf <- getVirtualFile uri
   case mvf of
-    Just (VirtualFile _ txt _) -> return $ Just (Rope.toText txt)
+    Just (VirtualFile _ txt) -> return $ Just (Rope.toText txt)
```
Comment (mpickering) [inline hie-plugin-api/Haskell/Ide/Engine/PluginUtils.hs]: We should make `VirtualFile` more abstract and provide an accessor to get the text out (I have this on my dejafu branch).

File: src/Haskell/Ide/Engine/Transport/LspStdio.hs [hunk 3]
```diff
@@ -219,8 +220,8 @@ mapFileFromVfs tn vtdi = do
   vfsFunc <- asksLspFuncs Core.getVirtualFileFunc
   mvf <- liftIO $ vfsFunc (J.toNormalizedUri uri)
   case (mvf, uriToFilePath uri) of
-    (Just (VFS.VirtualFile _ yitext _), Just fp) -> do
-      let text' = Rope.toString yitext
+    (Just (VFS.VirtualFile _ rope), Just fp) -> do
+      let text' = Rope.toString rope
```
Comment (mpickering) [inline src/Haskell/Ide/Engine/Transport/LspStdio.hs]: Same here.

File: src/Haskell/Ide/Engine/Transport/LspStdio.hs [hunk 4]
```diff
@@ -798,7 +799,7 @@ withDocumentContents reqId uri f = do
         (J.responseId reqId)
         J.InvalidRequest
         "Document was not open"
-    Just (VFS.VirtualFile _ txt _) -> f (Rope.toText txt)
+    Just (VFS.VirtualFile _ txt) -> f (Rope.toText txt)
```
Comment (mpickering) [inline src/Haskell/Ide/Engine/Transport/LspStdio.hs]: Same here..

File: src/Haskell/Ide/Engine/Transport/LspStdio.hs (hunk 1 above)
Comment (mpickering) [reply src/Haskell/Ide/Engine/Transport/LspStdio.hs]: I don't know either is this @bubba who changed it?

Comment (bubba) [reply src/Haskell/Ide/Engine/Transport/LspStdio.hs]: I have a branch for these options, I'll apply it on top of this after this is merged https://github.com/bubba/haskell-ide-engine/tree/new-haskell-lsp-options

**Comment by alanz** [comment]
Closed in favour of #1457

**Pull request closed without merging**